### PR TITLE
[WWST-4914]: implment DTH for scene keypad

### PR DESCRIPTION
**Pull request by FenMei007S**
@tpmanley  @greens plz review it

### CODE REVIEW

**Comment by FenMei007S** [comment]
This is for GDKES 4 and 6 buttons scene keypad, @greens  can u make 3 buttons scene keypad? another device is 3 buttons.what's more, the automation can only set the first button as trigger.

File: devicetypes/smartthings/zigbee-scene-keypad.src/zigbee-scene-keypad.groovy [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+def parseAttrMessage(description) {
+	def descMap = zigbee.parseDescriptionAsMap(description)
+    
+    	Map eventDescMap = zigbee.parseDescriptionAsMap(description)
+	log.debug "eventDescMap is $eventDescMap"
+    
+	def map = [:]
+	if (descMap?.clusterInt == 0x0017 || descMap?.clusterInt == 0xFE05) {
+	   def event = [:]
+       def buttonNumber
```
Comment (greens) [inline devicetypes/smartthings/zigbee-scene-keypad.src/zigbee-scene-keypad.groovy]: mixed spaces with tabs

File: devicetypes/smartthings/zigbee-scene-keypad.src/zigbee-scene-keypad.groovy [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+def refresh() {
+	return zigbee.enrollResponse()
+}
+
+def ping() {
+	refresh()
+}
+
+def configure() {
+	return configureReporting() + zigbee.enrollResponse()
```
Comment (greens) [inline devicetypes/smartthings/zigbee-scene-keypad.src/zigbee-scene-keypad.groovy]: `configureReporting()` is undefined.

Perhaps you meant to call `zigbee.configureReporting(...)` for the IAS cluster or something?

File: devicetypes/smartthings/zigbee-scene-keypad.src/zigbee-scene-keypad.groovy [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+ *
+ *	Author: f.mei@samsung.com
+ *	Date: 2019-02-18
+ */
+
+import groovy.json.JsonOutput
+import physicalgraph.zigbee.zcl.DataType
+
+metadata {
+	definition (name: "Zigbee Scene Keypad", namespace: "smartthings", author: "SmartThings",mcdSync: true) {
```
Comment (greens) [inline devicetypes/smartthings/zigbee-scene-keypad.src/zigbee-scene-keypad.groovy]: nit: space after "SmartThings"

File: devicetypes/smartthings/zigbee-scene-keypad.src/zigbee-scene-keypad.groovy [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+def initialize() {
+	
+}
+
+private addChildButtons(numberOfButtons) {
+	for(def i : 2..numberOfButtons) {
+		try {
+			String childDni = "${device.deviceNetworkId}:$i"
+			def componentLabel = (device.displayName.endsWith(' 1') ? device.displayName[0..-2] : device.displayName) + "${i}"
+			addChildDevice("Child Button", "${device.deviceNetworkId}:${i}", device.hubId,
```
Comment (greens) [inline devicetypes/smartthings/zigbee-scene-keypad.src/zigbee-scene-keypad.groovy:118]: I believe that `Child Button` supports held as well. You'll see "held" in the automation trigger list even though this device doesn't seem to generate "held" events.

File: devicetypes/smartthings/zigbee-scene-keypad.src/zigbee-scene-keypad.groovy [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+       event = createEvent(name: "button", value: "pushed", data: [buttonNumber: buttonNumber], descriptionText: "pushed", isStateChange: true)
+       if(buttonNumber != 1) {
+       		sendEventToChild(buttonNumber, event)
+       }else{
+       		sendEvent(event)
+       }
+    }
+}
+
+def getButtonEvent(descMap) {
```
Comment (greens) [inline devicetypes/smartthings/zigbee-scene-keypad.src/zigbee-scene-keypad.groovy]: This function is unused and should be removed

File: devicetypes/smartthings/zigbee-scene-keypad.src/zigbee-scene-keypad.groovy [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+ */
+
+import groovy.json.JsonOutput
+import physicalgraph.zigbee.zcl.DataType
+
+metadata {
+	definition (name: "Zigbee Scene Keypad", namespace: "smartthings", author: "SmartThings",mcdSync: true) {
+		capability "Actuator"
+		capability "Button"
+		capability "Holdable Button"
```
Comment (greens) [inline devicetypes/smartthings/zigbee-scene-keypad.src/zigbee-scene-keypad.groovy]: This button does not seem to be holdable.

Comment (greens) [reply devicetypes/smartthings/zigbee-scene-keypad.src/zigbee-scene-keypad.groovy]: i.e. does not generate "held" events

File: devicetypes/smartthings/zigbee-scene-keypad.src/zigbee-scene-keypad.groovy [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+}
+
+def configure() {
+	return configureReporting() + zigbee.enrollResponse()
+}
+
+def installed() {
+    def numberOfButtons = getChildCount()
+	sendEvent(name: "numberOfButtons", value: numberOfButtons, displayed: false)
+	sendEvent(name: "checkInterval", value: 2 * 60, displayed: false, data: [protocol: "zigbee", hubHardwareId: device.hub.hardwareID])
```
Comment (greens) [inline devicetypes/smartthings/zigbee-scene-keypad.src/zigbee-scene-keypad.groovy]: This check interval is way too short. Device Health will ping the device every minute. That's far too often.

I would recommend 32 minutes.

**Comment by greens** [comment]
@FenMei007S who made the metadata you're using in this DTH?

**Comment by FenMei007S** [comment]
> @FenMei007S who made the metadata you're using in this DTH?

Mr Donald Kirker made it

File: devicetypes/smartthings/zigbee-scene-keypad.src/zigbee-scene-keypad.groovy [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+		result = createEvent(descriptionText: "Wake up event came in", isStateChange: true)
+	}
+	log.debug "Description ${description} parsed to ${result}"
+	return result
+}
+
+def parseAttrMessage(description) {
+	def descMap = zigbee.parseDescriptionAsMap(description)
+    
+    	Map eventDescMap = zigbee.parseDescriptionAsMap(description)
```
Comment (greens) [inline devicetypes/smartthings/zigbee-scene-keypad.src/zigbee-scene-keypad.groovy]: this variable and line is redundant

File: devicetypes/smartthings/zigbee-scene-keypad.src/zigbee-scene-keypad.groovy (hunk 4 above)
Comment (greens) [reply devicetypes/smartthings/zigbee-scene-keypad.src/zigbee-scene-keypad.groovy:118]: I think, according to my understanding from @dkirker that you'll need to send a "supportedButtonValues" event for each child.

File: devicetypes/smartthings/zigbee-scene-keypad.src/zigbee-scene-keypad.groovy [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+		capability "Actuator"
+		capability "Button"
+		capability "Configuration"
+		capability "Refresh"
+		capability "Sensor"
+		capability "Health Check"
+
+		fingerprint profileId: "0104", inClusters: "0000, 0003, 0004, 0005", outClusters: "0003, 0004, 0005", manufacturer: "REXENSE", model: "HY0048", deviceJoinName: "情景开关 1", vid: "generic-4-button-alt"
+		fingerprint profileId: "0104", inClusters: "0000, 0003, 0004, 0005", outClusters: "0003, 0004, 0005", manufacturer: "REXENSE", model: "0106-G", deviceJoinName: "情景开关 1", vid: "generic-6-button-alt"
+    }
```
Comment (greens) [inline devicetypes/smartthings/zigbee-scene-keypad.src/zigbee-scene-keypad.groovy:31]: spaces and not tabs

File: devicetypes/smartthings/zigbee-scene-keypad.src/zigbee-scene-keypad.groovy (hunk 8 above)
Comment (greens) [reply devicetypes/smartthings/zigbee-scene-keypad.src/zigbee-scene-keypad.groovy]: 👀

**Review by greens — APPROVED** [review]

**Pull request merged**